sympy/tensor/array/sparse_ndim_array.py

```diff
@@ -111,19 +111,40 @@ def reshape(self, *newshape):
 
         return type(self)(*(newshape + (self._sparse_array,)))
 
-    def _eval_derivative_n_times(self, s, n):
-        from sympy import Integer, Derivative
-        if isinstance(n, (int, Integer)):
+    def _eval_derivative_array(self, arg):
+        from sympy.core.compatibility import Iterable
+        from sympy.matrices.common import MatrixCommon
+        from sympy import Derivative
+        from sympy import Dict
+        if isinstance(arg, (Iterable, Tuple, MatrixCommon, NDimArray)):
+            return type(self)([self.diff(x) for x in arg])
```

Comment:
This case still cast sparse array to dense array. I have tried `derive_by_array` but `derive_by_array` itself has the same problem. Maybe an algorithm to calculate the new index after concatenating some sparse arrays, so that we can initiate a sparse array with the passed dictionary. That is to say:

[self.diff(x) for x in arg] -> [a, b, c] where a, b, c are sparse arrays. 

The question is how to calculate the index of non-zero value of a, b, c in a new dictionary, say D, so that D can represent [a, b, c].